main/lsp/requests/completion.cc

```diff
@@ -7,50 +7,122 @@ using namespace std;
-                                                                          core::TypePtr receiver, string_view name) {
-    UnorderedMap<core::NameRef, vector<core::SymbolRef>> result;
-    typecase(
-        receiver.get(),
-        [&](core::ClassType *c) {
-            const auto &owner = c->symbol.data(gs);
-            for (auto member : owner->membersStableOrderSlow(gs)) {
-                auto sym = member.second;
-                if (sym.data(gs)->isMethod() && hasSimilarName(gs, sym.data(gs)->name, name)) {
-                    result[sym.data(gs)->name].emplace_back(sym);
-                }
-            }
-            for (auto mixin : owner->mixins()) {
-                result = mergeMaps(std::move(result),
-                                   findSimilarMethodsIn(gs, core::make_type<core::ClassType>(mixin), name));
+// First of pair is "found at this depth in the ancestor hierarchy"
+// Second of pair is method symbol found at that depth, with name similar to prefix.
+vector<SimilarMethod> similarMethodsForClass(const core::GlobalState &gs, core::SymbolRef receiver,
+                                             string_view prefix) {
+    auto result = vector<SimilarMethod>{};
+
+    int depth = -1;
+    for (auto ancestor : ancestors(gs, receiver)) {
+        depth++;
+        for (auto [memberName, memberSymbol] : ancestor.data(gs)->members()) {
+            if (!memberSymbol.data(gs)->isMethod()) {
+                continue;
             }
-            if (owner->superClass().exists()) {
-                result =
-                    mergeMaps(std::move(result),
-                              findSimilarMethodsIn(gs, core::make_type<core::ClassType>(owner->superClass()), name));
+
+            if (hasSimilarName(gs, memberName, prefix)) {
+                result.emplace_back(SimilarMethod{depth, receiver, memberSymbol});
             }
+        }
+    }
+
+    return result;
+}
+
+vector<SimilarMethod> mergeSimilarMethods(vector<SimilarMethod> &&left, vector<SimilarMethod> &&right) {
+    left.insert(left.end(), right.begin(), right.end());
+    return std::move(left);
+}
+
+vector<SimilarMethod> similarMethodsForReceiver(const core::GlobalState &gs, const core::TypePtr receiver,
+                                                string_view prefix) {
+    auto result = vector<SimilarMethod>{};
+
+    typecase(
+        receiver.get(), [&](core::ClassType *type) { result = similarMethodsForClass(gs, type->symbol, prefix); },
+        [&](core::AppliedType *type) { result = similarMethodsForClass(gs, type->klass, prefix); },
+        [&](core::AndType *type) {
+            // We take the union here rather than take the intersection. (Better to suggest a method that someone
+            // is looking for and then give a type error, rather than have them sit wondering why it's missing.)
+            result = mergeSimilarMethods(similarMethodsForReceiver(gs, type->left, prefix),
```

Comment:
Rather than merge things here, why not pass in `result` as an out parameter?

`similarMethodsForReceiver` can take in a reference to a vector and mutate it.

e.g.:

```c++
void similarMethodsForReceiver(const core::GlobalState &gs, const core::TypePtr receiver,
                                                string_view prefix, vector<SimilarMethod> &result);

similarMethodsForReceiver(gs, type->left, prefix, result);
```

Reply:
- one less argument in the parameter list means fewer things to think about
- dmitry has drilled into me that out parameters are usually harder to follow (turn local reasoning into global reasoning + mutation)
- i've been taught to write dumb, slow, understandable code until it becomes clear that it's the bottleneck. we currently believe that populating GlobalState (from both the fast and slow paths) dominates this computation, because these completion lists will usually be tiny.

Reply:
👍